notebooks/implicit_differentiation.ipynb

```diff
@@ -0,0 +1,561 @@
+    "    \n",
+    "    # wrap force_fn with a lax.stop_gradient to prevent a CustomVJPException\n",
+    "    no_grad_force_fn = jit(lambda x: lax.stop_gradient(force_fn(x)))\n",
+    "    \n",
+    "    # make the dependence on the variables we want to differentiate explicit\n",
+    "    explicit_force_fn = jit(lambda R, p: force_fn(R, **p))\n",
+    "    \n",
+    "    def solver(params, x):\n",
+    "        # params are unused\n",
+    "        del params\n",
+    "        # need to use no_grad_force_fn!\n",
+    "        return run_minimization_while(no_grad_force_fn, x, shift, dt_start=0.001, dt_max=0.005)[0]\n",
+    "    \n",
+    "    decorated_solver = jaxopt.implicit_diff.custom_root(explicit_force_fn)(solver)\n",
+    "    \n",
+    "    R_final = decorated_solver(None,R_init)\n",
+    "    \n",
+    "    # Here we can just use our original energy_fn/force_fn\n",
+    "    return energy_fn(R_final), jnp.amax(jnp.abs(force_fn(R_final)))\n",
+    "\n",
+    "(imp_e, imp_f), imp_g = jax.value_and_grad(implicit_diff,has_aux=True)(param_dict,R_init,displacement)\n",
+    "print('Energy        : ',imp_e)\n",
+    "print('Max_grad_force: ',imp_f)\n",
+    "print('Gradient of the energy:')\n",
+    "print(imp_g['sigma'][0])\n",
+    "print(imp_g['alpha'][0])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "biblical-pharmacy",
+   "metadata": {},
+   "source": [
+    "As you can see we do not have to change anything compared to our explicit version when we want to compute gradients with respect to more then one parameter."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "dried-shame",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# As you can see the we get the same results.\n",
+    "print(jax.tree_map(jnp.allclose,exp_e,imp_e))\n",
+    "print(jax.tree_map(jnp.allclose,exp_f,imp_f))\n",
+    "print(jax.tree_map(jnp.allclose,exp_g,imp_g))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "happy-aviation",
+   "metadata": {},
+   "source": [
+    "We use [jaxopt](https://github.com/google/jaxopt) to define implicit gradients for our energy minimization routine. To this end we need to define two new force functions from our original force_fn. This is necessary for two seperate reasons. \n",
+    "\n",
+    "For one [jaxopt.implicit_diff.custom root](https://jaxopt.github.io/stable/_autosummary/jaxopt.implicit_diff.custom_root.html#jaxopt.implicit_diff.custom_root) requires a function that takes the parameter that we want to differentiate as an explicit input. To this end we define a new explicit_force_fn as\n",
+    "```python \n",
+    "explicit_force_fn = jit(lambda R, p: force_fn(R,**p))\n",
+    "```\n",
+    "\n",
+    "Furthermore we cannot pass our original force_fn to our solver as this will cause a ```CustomVJPException```. We can just wrap the output of our ```force_fn``` with ```lax.stop_gradient``` in order to fix this issue.\n",
```

Comment:
Do you understand why this exception comes up?

Reply:
No. I'm going to raise an issue about that with jaxopt.

Reply:
Awesome, thanks!

Reply:
A small correction: You get a memory leak + a CustomVJPException.
This memory leak show's up because I'm using a python while loop. If you'd replace that with a lax.while_loop then you only get the CustomVJPException. I forgot to mention that I opened up  jaxopt issue 111 because of that. I believe the CustomVJPException is caused by the way the energy functions are constructed, but I'm not completely sure about that.